### PR TITLE
Turn dask to an optional requirement

### LINKED ISSUE #3410: Request: make dask[array] an optional dependency
In #2494, `dask` was made an optional dependency; in #2572, it was removed from `requirements.txt` (following #2571).

It looks as thought for version 0.14.0, `dask` is back to being a [required dependency](https://github.com/scikit-image/scikit-image/blob/master/requirements/default.txt) again. Was this an oversight when the `requirements.txt` file was refactored, or is `dask` now considered to be a required dependency of `scikit-image`?

(Background and motivation: Enthought packages `scikit-image` for EDM and Canopy, and the `dask` dependency turns `scikit-image` into quite a heavyweight package; if we can reasonably declare that `dask` is an optional dependency, we'd be quite happy to do so. :-)

### PR DESCRIPTION
**Pull request by sciunto**
## Description

WIP: have to handle the warnings in the tests...

Fixes #3410 

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](https://scikit-image.org/docs/dev/contribute.html)]

## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by soupault** [comment]
>WIP: have to handle the warnings in the tests...

There was a discussion on the warnings in https://github.com/scikit-image/scikit-image/pull/2647, if that helps.

**Comment by hmaarrfk** [comment]
should these be hard failures? If I set numworkers to something other than 1 and it gives me 1 worker, I might want things to fail no?

**Comment by jni** [comment]
@hmaarrfk I'm unclear about this. I might actually prefer to be allowed to get on with my life if I'm in an interactive session, then make a mental note to install dask for future work.

I might say that I would prefer to do this optional dependency thing only once we have `pip install scikit-image[complete]` working. (Then the warning can mention that, as well as an equivalent conda meta-package.) Others might disagree and I'm happy to be overruled.

**Comment by stefanv** [comment]
It's an easy enough solution that we may just as well push it now for
the two or three places it is being used.  Who knows when or if
`scikit-image[all]` will become a thing, but we can figure it out then.

I'd argue not to fail if multiple CPUs are requested but only one is
available, because we want skimage code to run unmodified across
different machines, with any valid configuration of dependencies.  But a
warning, in that case, would be required.

**Comment by hmaarrfk** [comment]
Makes sense guys.
I guess I can make a warningsfilter to get the behavior I want.

**Comment by sciunto** [comment]
@scikit-image/core it's ready for a final review.

File: skimage/restoration/_cycle_spin.py [hunk 1]
```diff
@@ -143,8 +148,11 @@ def _run_one_shift(shift):
         tmp = func(xs, **func_kw)
         return _roll_axes(tmp, -np.asarray(shift))
 
+    if not dask_available:
+        warn('The optional dask dependency is not installed. '
+             'The number of worker is set to 1.')
```
Comment (jni) [inline skimage/restoration/_cycle_spin.py]: imho this warning should only be triggered *if* dask is not available *and* the number of workers is set to a value other than 1. Which, incidentally, you're not setting here! So I think you need to

- change the condition to also check whether `num_workers` is set to `>1` or `None`
- set `num_workers = 1` inside the condition
- add a test for all three conditions, including a test for `num_workers=1` *not* raising a warning.

Comment (sciunto) [reply skimage/restoration/_cycle_spin.py]: oh, that's what I wanted to do! I forgot half of the condition... :( I'll fix it.

**Comment by sciunto** [comment]
I've fixed the condition.

File: skimage/restoration/_cycle_spin.py (hunk 1 above)
Comment (jni) [reply skimage/restoration/_cycle_spin.py]: @sciunto 

a) "number of worker*s* is set to 1"
b) could you add a test to verify that `num_workers=1` results in no warning being raised
c) could you expand your message to say, "to silence this warning, install dask or explicitly set `num_workers=1` when calling the `cycle_spin` function"?

Thanks!

Comment (sciunto) [reply skimage/restoration/_cycle_spin.py]: I think all your requests are satisfied. I don't think we need an extra test.

**Comment by emmanuelle** [comment]
Merging, thanks.

**Comment by emmanuelle** [comment]
Merging, thanks.

**Pull request merged**